util/messaging.py

```diff
@@ -0,0 +1,234 @@
+        If 'job' and 'task' address the local node, the call is performed locally.
+        'job' - Job name of the local cluster node
+        'task' - Task index of the local cluster node
+        'function' - Name of the remote function (member of class derived from ClusterMessagingClient)
+        'callback' - Will be called with return value tuple of remote function as arguments,
+            if RPC call succeeded. Omitting it (None) results in dependency free state
+            (the local client can safely be stopped/ended).
+            Can also be an Event object (used internally by 'call').
+        'args' - Arguments to the remote function.
+        Returns a callback id (0 in case of no callback).
+        '''
+        # target client id (shared_name for the apropos queue)
+        id = self._id_pattern % (job, task)
+        if self.id == id:
+            # local call - no queueing needed
+            log.step('Local call of function: %s, arguments: %r' % (function, args))
+            results = self._call(function, args)
+            if callback:
+                callback(results)
+            return -1
+        log.step('Calling worker: %s, function: %s, arguments: %r' % (id, function, args))
+        # default value in case of no callback, if not waiting for responseno
+        # and no response should be produced on the receiving end.
+        response_id = 0
+        if callback:
+            with self._response_id_lock:
+                # acquire a new response id
+                response_id = self._response_id_counter = self._response_id_counter + 1
+                # register callback or Event object as 'waiting for response'
+                self._waiting_for_response[response_id] = callback
+        # en-pickle header and call arguments
+        content = pickle.dumps((self.id, False, response_id, function, args))
+        # get enqueue op of outbound queue of the targeted client
+        enqueue = self._outbound_enqueues[id]
+        # enqueue locally for sending
+        self._to_send.put((enqueue, { self._ph_content: content }))
+        # passing back response_id (used for synchronous RPCs - see 'call')
+        return response_id
+
+    def call(self, job, task, function, *args):
+        '''
+        Does a synchronous (blocking) RPC call to a remote node's function.
+        If 'job' and 'task' address the local client, the call is performed locally.
+        'job' - Job name of the local cluster node
+        'task' - Task index of the local cluster node
+        'function' - Name of the remote function (member of class derived from ClusterMessagingClient)
+        'args' - Arguments to the remote function.
+        '''
+        if job == self.job and task == self.task:
+            # local call - no queueing needed
+            log.step('Local call of function: %s, arguments: %r' % (function, args))
+            return self._call(function, args)
```

Comment:
Local will not be blocking, will it ?

Reply:
It will, as it is the direct call to the actual/underlying routine.